### DISCUSSION TITLE
Nested GraphQL Mutations

### DISCUSSION THREAD

**Discussion by leebyron** [graphql/graphql-wg#815] [Ideas]
Please see these links by @aleksandarsusnjar 

   - Details document: [GraphQL Mutations and Large Schemas](https://github.com/aleksandarsusnjar/graphql-proposals/blob/main/graphql-mutations-and-large-schemas.md)
   - [Nested mutations (PowerPoint Show)](https://github.com/aleksandarsusnjar/graphql-proposals/blob/main/Nested%20mutations.ppsx) (DRAFT before the meeting)

This discussion thread is for capturing feedback and ideas about this topic

**Comment by xuorig** [comment]
Just as a precision I think it's important to highlight that I don't think this statement is factual:

> This does not require the spec to change, it is not going against it:

From the [specification](https://spec.graphql.org/June2018/#sec-Normal-and-Serial-Execution):

> the resolution of fields other than top‐level mutation fields must always be side effect‐free and idempotent, the execution order must not affect the result

So allowing nested mutation would in fact require a spec change ☝️  to allow side effects to occur on any field (or certain fields depending on the proposal).

**Reply by aleksandarsusnjar** [reply]
@xuorig I may have missed some things but note that that many treat this as a "soft" requirement:

- Absolutely everything, including queries, have side-effects in some systems. Think (audit) logging, resource use, etc.
- Different execution orders may yield the same result - when they are independent.
- Who is responsible for "the execution order must not affect the result"? This is on client's shoulders' as much as servers'. In a highly used system the same execution order may yield different results at different times due to concurrent requests from other (or even the same client).
- If the server does implement sequential execution, which I believe many do anyway (e.g. because it is simpler or to control resource utilization by single clients or to be able to reuse the same database connection), the execution order will be deterministic anyway.

So, yes, the spec could use some cleanup, but nothing major is _required_.

**Reply by aleksandarsusnjar** [reply]
Note the context of that:

> Normally the executor can execute the entries in a grouped field set in whatever order it chooses (normally in parallel). Because the resolution of fields other than top‐level mutation fields must always be side effect‐free and idempotent, the execution order must not affect the result, and hence the server has the freedom to execute the field entries in whatever order it deems optimal.

... specifically: "and hence the server has the freedom to execute the field entries in whatever order it deems optimal."

... so it is very "soft" and can be interpreted as "the client must carefully split requests if it wants to ensure any order".

**Reply by xuorig** [reply]
> Absolutely everything, including queries, have side-effects in some systems. Think (audit) logging, resource use, etc.

This is not what is intended by side-effect. A side-effect to the underlying data-system is the wording from the spec. It's fair to assume here we're talking about effects that modify the system in a way that could affect subsequent queries results. Not logging or resource use. 

Non top-level Mutation fields **must** be **idempotent** in the current version of the spec. Your proposal changes that. It doesn't make the proposal a bad idea, but it does require changing the spec, in a **major** way.

> Different execution orders may yield the same result - when they are independent.

They may, but they **must** not. This is a big difference.

Example: A GraphQL gateway could automatically retry execution on certain fields when they return a transient error, because it can be certain the field is idempotent. If we remove this part of the spec, this is now quite dangerous behavior.

**Reply by aleksandarsusnjar** [reply]
> They may, but they must not. This is a big difference.

I think you meant "... but they MAY not". I am not saying that there aren't the cases to the contrary. What I am saying is that there are many perfectly valid cases for which this is an unnecessary obstacle. There is a huge negative impact of not allowing nested mutations and it does not have to exist.

**Reply by aleksandarsusnjar** [reply]
Also do note that the client absolutely *has* to be aware of the importance of order, at all times, with the current spec or any modification of it. Current spec wording and/or the strictest interpretation of it do not save the client from that responsibility. They must serialize their requests accordingly and not execute them concurrently, for example (which they could).

**Reply by xuorig** [reply]
> I think you meant "... but they MAY not". I am not saying that there aren't the cases to the contrary. What I am saying is that there are many perfectly valid cases for which this is an unnecessary obstacle. There is a huge negative impact of not allowing nested mutations and it does not have to exist.

Again, I hear your case, what I'm saying is, it does require a spec change. And changing non top-level fields from being idempotent to non-idempotent is a major one on my view. That is my only point here 👍

**Reply by aleksandarsusnjar** [reply]
Let's agree to that. Changing the spec is highly desirable in either case.

**Reply by xuorig** [reply]
👍  the reason I think it's important to highlight is that backward compatibility and favouring no change are two [guiding principles](https://github.com/graphql/graphql-spec/blob/main/CONTRIBUTING.md#guiding-principles). That does not mean the proposal is not great, it just means it needs a "higher bar of added value".

**Reply by aleksandarsusnjar** [reply]
Yep. I paid attention to these. And, yes, I'm quite aware that this needs to pass a higher bar. The added value is undeniably there, as evidenced by existing content/implementations. To avoid this taking forever I can see this being addressed in potentially a sequence of multiple "evolutions":

1. Spec wording and guidance improvements. Allow with due caution.
2. Evolve the spec allow better control of the execution sequence, in some way - I listed multiple possibilities which can build on one another.

**Comment by aleksandarsusnjar** [comment]
> **6.3.1 Normal and Serial Execution**
> Normally the executor can execute the entries in a grouped field set in whatever order it chooses (normally in parallel). Because the resolution of fields other than top-level mutation fields must always be side effect-free and idempotent, the execution order must not affect the result, and hence the service has the freedom to execute the field entries in whatever order it deems optimal.

Could be update to something like:

> Normally the executor can execute the entries in a grouped field set in whatever order it chooses (normally in parallel). Because the the order of resolution of fields other than top-level mutation fields is not specified by default, the execution of nested fields must either be side effect-free or otherwise idempotent - the client may not expect any particular execution order and must not be affect by the result. Hence the service has the freedom to execute the field entries in whatever order it deems optimal.

Note: 
1.  Existing servers without nested mutations will not be affected by this precisely because they do not have nested mutations (i.e. nested field execution is side-effect free in the sense implied in the spec).
2. Clients of those servers (1) will not be affected because they aren't using nested mutations (because they don't exist).
3. Existing servers implementing nested mutations already ignored the present spec guidance and would be made more or entirely valid with a change as above.
4. Clients of those servers (3) are already made with corresponding expectations. Nothing would change.

A follow up change, sooner or later, may add a description of how to specify server abilities and behaviour that is non-default (i.e. not as described in the that paragraph above) and how can clients properly leverage that functionality.
